packages/react/src/components/sideNavigation/mainLevel/MainLevel.tsx

```diff
@@ -0,0 +1,198 @@
+import xor from 'lodash.xor';
+import React, { cloneElement, isValidElement, ReactNode, useContext } from 'react';
+
+import classNames from '../../../utils/classNames';
+import styles from './MainLevel.module.scss';
+import SideNavigationContext from '../SideNavigationContext';
+import { FCWithName } from '../../../common/types';
+import { IconAngleDown, IconLinkExternal } from '../../../icons';
+
+export type MainLevelProps = {
+  /**
+   * If `true`, the item will be marked as active
+   */
+  active?: boolean;
+  /**
+   * Additional class names to apply to the side navigation main level
+   */
+  className?: string;
+  /**
+   * The id of the side navigation main level.
+   */
+  id: string;
+  /**
+   * href attribute of the side navigation main level
+   */
+  href?: string;
+  /**
+   * Optional icon of the side navigation main level.
+   */
+  icon?: React.ReactNode;
+  /**
+   * Index of the side navigation main level.
+   */
+  index?: number;
+  /**
+   * Label of the side navigation main level.
+   */
+  label: string;
+  /**
+   * A border to distinct main level item from previous item.
+   */
+  withDivider?: boolean;
+  /**
+   * Boolean indicating whether the main level link will lead user to external domain.
+   */
+  external?: boolean;
+  /**
+   * Aria label for opening main level link in an external domain
+   */
+  openInExternalDomainAriaLabel?: string;
+  /**
+   * Override or extend the styles applied to the component
+   */
+  style?: React.CSSProperties;
+  /**
+   * Override or extend the styles applied to the component
+   */
+  children?: ReactNode;
+  /**
+   * Callback function fired when the tag is clicked. If set, the tag will be clickable.
+   */
+  onClick?: (
+    event:
+      | React.MouseEvent<HTMLAnchorElement, MouseEvent>
+      | React.MouseEvent<HTMLButtonElement, MouseEvent>
+      | React.KeyboardEvent<HTMLAnchorElement>
+      | React.KeyboardEvent<HTMLButtonElement>,
+  ) => void;
+};
+
+const LeftIcon = ({ icon }: { icon: React.ReactNode }) => (
+  <span className={styles.leftIcon} aria-hidden>
+    {icon}
+  </span>
+);
+
+const RightIcon = ({ icon, className = '' }: { icon: React.ReactNode; className?: string }) => (
+  <span className={classNames(styles.rightIcon, className)} aria-hidden>
+    {icon}
+  </span>
+);
+
+const Label = ({ label }: { label: string }) => <span className={styles.label}>{label}</span>;
+
+export const MainLevel = ({
+  active,
+  children,
+  className,
+  href,
+  icon,
+  id,
+  index,
+  label,
+  withDivider,
+  external,
+  openInExternalDomainAriaLabel,
+  onClick,
+  style,
+  ...rest
+}: MainLevelProps) => {
+  const menuId = `${id}-menu`;
+  const { autoCollapseOthers, openMainLevels, setMobileMenuOpen, setOpenMainLevels } = useContext(
+    SideNavigationContext,
+  );
+  const activeRef = React.useRef<boolean>(active);
+  const open = openMainLevels.includes(index as number);
+
+  const subLevels = React.Children.map(children, (child) => {
```

Comment:
I wonder if this cloning could be avoided by providing context to subLevels? Not sure if it would just require more work? Also it might have bring additional renderings. Is there any risks related to cloning? If user provides a ref inside the sublevel, and we clone it here, then does the original ref still work?

Reply:
Apparently the ref from the original component will be preserved: https://reactjs.org/docs/react-api.html#cloneelement